File: apps/front/src/api/defi.ts
import { createWrapper, FetchPolicy } from '@api';
import { coreApiFetch as apiFetch } from '@api';

// --------------------
//  MODELS DEFINITION
// --------------------

export interface DefiAnswer {
	id: string;
	text: string;
	leafReward: number;
	percentReward: number;
}

export interface DefiQuestion {
	id: string;
	text: string;
	responses: DefiAnswer[];
}

export interface DailyDefi {
	id: string;
	defi: string;
	category: string;
	leafReward: number;
	status: 'PENDING' | 'COMPLETED';
	earned: number;
	overQuestions?: DefiQuestion;
}

export interface CompleteDefiResponse {
	status: string;
	reward: number;
}

// --------------------
//    API DEFINITION
// --------------------

const getDailyDefisApi = () =>
	apiFetch<DailyDefi[]>('/defi/daily', {
		method: 'GET',
		auth: true,
	});

export const completeDefi = (defiId: string, answerId: string) =>
	apiFetch<CompleteDefiResponse>('/defi/complete', {
		method: 'POST',
		body: JSON.stringify({ defiId, answerId }),
		auth: true,
	});

// --------------------
//  WRAPPER DEFINITION
// --------------------

export const dailyDefiWrapper = createWrapper<DailyDefi[], void>({
	apiCall: getDailyDefisApi,
	cacheKey: () => `daily_defis_${new Date().toISOString().split('T')[0]}`,
	policy: {
		policy: FetchPolicy.stale_while_revalidate,
		staleTtlMs: 5 * 60 * 1000,
	},
});

File: apps/core/internal/models/defi.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type DefiDefinition struct {
	Id            string        `json:"id"`
	Defi          string        `json:"defi"`
	Category      string        `json:"category"`
	LeafReward    int           `json:"leafReward"`
	PercentReward int           `json:"percentReward"`
	OverQuestions *DefiQuestion `json:"overQuestions,omitempty"`
}

type DefiQuestion struct {
	Id        string       `json:"id"`
	Text      string       `json:"text"`
	Responses []DefiAnswer `json:"responses"`
}

type DefiAnswer struct {
	Id            string `json:"id"`
	Text          string `json:"text"`
	LeafReward    int    `json:"leafReward"`
	PercentReward int    `json:"percentReward"`
}

type DailyDefi struct {
	gorm.Model
	UserId       uint      `gorm:"primaryKey"`
	DefiId       string    `gorm:"type:varchar(50)"`
	DateAssigned time.Time `gorm:"type:date;index:idx_user_date"`
	Status       string    `gorm:"type:varchar(20);default:'PENDING'"`
	RewardEarned int       `gorm:"default:0"`
}
